### PR TITLE
Text norm

### PR DESCRIPTION
**Pull request by yzhang123**
* text normalization in WFST
* merge electronic (Evelina)
* updated docs for ITN and TN
* updated Tutorials
* unified deployment route
* unit tests for both

### CODE REVIEW

File: nemo_text_processing/inverse_text_normalization/inverse_normalize.py [hunk 1]
```diff
@@ -13,216 +13,201 @@
...
+    def inverse_normalize_list(self, texts: List[str], verbose=False) -> List[str]:
         """
-        Generates permutations of string serializations of given dictionary
+        NeMo inverse text normalizer 
 
         Args:
-            tokens: list of dictionaries
-            prefix: prefix string
-            idx:    index of next dictionary
+            texts: list of input strings
```
Comment (ekmb) [inline nemo_text_processing/inverse_text_normalization/inverse_normalize.py:49]: verbose arg is missing in the docstirng

File: nemo_text_processing/inverse_text_normalization/inverse_normalize.py [hunk 2]
```diff
@@ -13,216 +13,201 @@
...
+
+    def inverse_normalize(self, text: str, verbose: bool) -> str:
+        """
+        Main function. Inverse normalizes tokens from spoken to written form
+            e.g. twelve kilograms -> 12 kg
 
-    Args:
-        tagged_text: tagged text
+        Args:
+            text: string that may include semiotic classes.
```
Comment (ekmb) [inline nemo_text_processing/inverse_text_normalization/inverse_normalize.py:68]: verbose is missing

File: nemo_text_processing/inverse_text_normalization/inverse_normalize.py [hunk 3]
```diff
@@ -13,216 +13,201 @@
...
-        if verbose:
-            print(output)
-        return output
-    raise ValueError()
+    def select_tag(self, lattice: 'pynini.FstLike') -> str:
+        """
+        Given tagged lattice return shortest path
 
+        Args:
+            tagged_text: tagged text
```
Comment (ekmb) [inline nemo_text_processing/inverse_text_normalization/inverse_normalize.py:170]: docstring doesn't match arg name

File: nemo_text_processing/text_normalization/graph_utils.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/graph_utils.py:1]: Can this file be shared between ITN and TN?

File: nemo_text_processing/text_normalization/normalize.py [hunk 5]
```diff
@@ -12,198 +12,211 @@
...
-        res.append(text)
-    return res
-
-
-normalizers = {"identity": normalize_identity, "nemo": normalize_nemo}
+        input_case: expected input capitalization
+    """
+
+    def __init__(self, input_case: str):
+        assert input_case in ["lower_cased", "cased"]
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/normalize.py:43]: let's raise an error

File: nemo_text_processing/text_normalization/normalize.py [hunk 6]
```diff
@@ -12,198 +12,211 @@
...
+        self.tagger = ClassifyFst(input_case=input_case)
+        self.verbalizer = VerbalizeFinalFst()
+        self.parser = TokenParser()
+
+    def normalize_list(self, texts: List[str], verbose=False) -> List[str]:
+        """
+        NeMo text normalizer 
+
+        Args:
+            texts: list of input strings
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/normalize.py:54]: verbose is missing

File: nemo_text_processing/text_normalization/normalize.py [hunk 7]
```diff
@@ -12,198 +12,211 @@
...
+            res.append(text)
+        return res
+
+    def normalize(self, text: str, verbose: bool) -> str:
+        """
+        Main function. Normalizes tokens from written to spoken form
+            e.g. 12 kg -> twelve kilograms
+
+        Args:
+            text: string that may include semiotic classes.
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/normalize.py:74]: verbose is missing

File: nemo_text_processing/text_normalization/normalize.py [hunk 8]
```diff
@@ -12,198 +12,211 @@
...
+            tagged_text = pynini.escape(tagged_text)
+            verbalizer_lattice = self.find_verbalizer(tagged_text)
+            if verbalizer_lattice.num_states() == 0:
+                continue
+            output = self.select_verbalizer(verbalizer_lattice)
+            return output
+        raise ValueError()
+
+    def _permute(self, d: OrderedDict) -> List[str]:
+        """
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/normalize.py:101]: can this func and most of others in this file  be reused in ITN/TN?

File: nemo_text_processing/inverse_text_normalization/taggers/telephone.py [hunk 9]
```diff
@@ -17,9 +17,9 @@
 
 class TelefoneFst(GraphFst):
     """
-    Finite state transducer for classifying telephone
+    Finite state transducer for classifying telephone numbers
     """
 
     def __init__(self):
         super().__init__(name="telefone", kind="classify")
```
Comment (ekmb) [inline nemo_text_processing/inverse_text_normalization/taggers/telephone.py:24]: `telefone` -> `telephone` as in the Google dataset

File: nemo_text_processing/text_normalization/data_loader_utils.py [hunk 10]
```diff
@@ -105,18 +115,24 @@ def clean_generic(text: str) -> str:
...
     """
     acc = 0
     nums = len(preds)
     for i in range(nums):
         pred_norm = clean_generic(preds[i])
         label_norm = clean_generic(labels[i])
+
+        # print(f"inpu: {json.dumps(input[i])}")
+        # print(f"gold: {json.dumps(label_norm)}")
+        # print(f"pred: {json.dumps(pred_norm)}")
```
Comment (ekmb) [inline nemo_text_processing/text_normalization/data_loader_utils.py:135]: is this needed?

**Review by ekmb — CHANGES_REQUESTED** [review]

File: nemo_text_processing/text_normalization/data_loader_utils.py (hunk 10 above)
Comment (yzhang123) [reply nemo_text_processing/text_normalization/data_loader_utils.py:135]: deleted

File: docs/source/nemo_text_processing/text_normalization.rst [hunk 11]
```diff
@@ -8,12 +8,43 @@ For example,
 `"at 10:00"` -> `"at ten o'clock"` 
 and `"it weighs 10kg"` -> `"it weights ten kilograms"`.
 
-This tool is currently based on Pyhon Regex.
+NeMo Text Normalization is based on WFST-grammars :cite:`textprocessing-norm-zhang2021nemo`. We also provide a deployment route to C++ using Sparrowhawk -- an open-source version of Google Kestrel :cite:`textprocessing-norm-ebden2015kestrel`.
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply docs/source/nemo_text_processing/text_normalization.rst:11]: ok

File: nemo_text_processing/text_normalization/graph_utils.py (hunk 4 above)
Comment (yzhang123) [reply nemo_text_processing/text_normalization/graph_utils.py:1]: I tried and its not trivial, imports mess up references. Can i try this in another PR?

File: nemo_text_processing/text_normalization/normalize.py (hunk 8 above)
Comment (yzhang123) [reply nemo_text_processing/text_normalization/normalize.py:103]: I tried and its not trivial, imports mess up references. Can i try this in another PR?

File: nemo_text_processing/text_normalization/graph_utils.py (hunk 4 above)
Comment (ekmb) [reply nemo_text_processing/text_normalization/graph_utils.py:1]: Ok, let's leave it for a separate PR

**Review by ekmb — APPROVED** [review]

File: tools/ctc_segmentation/scripts/prepare_data.py [hunk 12]
```diff
@@ -62,6 +62,8 @@
     help='Set to True to use NeMo Normalization tool to convert numbers from written to spoken format.',
 )
 
+normalizer = Normalizer(input_case='lower_cased')
```
Comment (ekmb) [inline tools/ctc_segmentation/scripts/prepare_data.py:65]: please preserve the case here

**Review by ekmb — CHANGES_REQUESTED** [review]

**Review by ekmb — APPROVED** [review]

**Pull request merged**